staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_patch.go

```diff
@@ -0,0 +1,63 @@
+package cacher
+
+import (
+	"fmt"
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+)
+
+type clusterNameAndNsGetter interface {
+	GetClusterName() string
+	GetNamespace() string
+}
+
+// recalculateWatchCacheKey adjusts the key under which the given object will be stored in the watch cache.
+//
+// resources in the db are stored without the clusterName, since the reflector used by the cache uses the logicalcluster.Wildcard
+// the clusterName will be assigned to object by the storage layer upon retrieval.
+// we need take it into consideration and change the key to contain the clusterName
+// because this is how clients are going to be retrieving data from the cache.
+func recalculateWatchCacheKey(existingKey string, object runtime.Object) string {
+	var clusterName, ns string
+	key := existingKey
+
+	switch t := object.(type) {
+	case metav1.ObjectMetaAccessor:
+		clusterName = t.GetObjectMeta().GetClusterName()
+		ns = t.GetObjectMeta().GetNamespace()
+	case clusterNameAndNsGetter:
```

Comment:
What fulfills this interface and not ObjectMetaAccessor?

Reply:
by `unstructured.Unstructured`

Reply:
Maybe worth a comment, I am surprised that it's true